### PR TITLE
[JUJU-3648] Controller model table schema

### PR DESCRIPTION
**Pull request by SimonRichardson**
This patch is broken off from #15539. This defines the model table schema. This schema is intended to be small in scope. The current sole responsibility of this table is to ensure that the model exists and can therefore be opened by the db accessor worker.

Potential future work could include the number of accesses and last usage, but for now we purely just want to restrict open/view access. If the value is within the table, consider it fair game for access.

Additionally, I've moved the schema outside of the database package and into the domain package. This feels more like a logical location.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test --build-agent
```

### CODE REVIEW

File: domain/schema/controller_test.go [hunk 1]
```diff
@@ -75,6 +75,9 @@ func (s *schemaSuite) TestDDLApply(c *gc.C) {
 		"external_controller",
 		"external_controller_address",
 		"external_model",
+
+		// Model list
```
Comment (manadart) [inline domain/schema/controller_test.go:79]: Not sure we need this...

File: domain/schema/controller.go [hunk 2]
```diff
@@ -272,3 +273,16 @@ BEGIN
 END;
 `[1:]
 }
+
+func modelListSchema() string {
+	return `
+CREATE TABLE model_list (
+    uuid        TEXT PRIMARY KEY,
```
Comment (manadart) [inline domain/schema/controller.go:280]: I don't think we need the synthetic UUID, when we already have an entity identified by one.
Just have `uuid`, implying the model UUID that we use today.
I also wonder if we might need the model name here in the controller DB.

**Review by manadart — COMMENTED** [review]
Suggested change.

File: domain/schema/controller.go (hunk 2 above)
Comment (SimonRichardson) [reply domain/schema/controller.go:280]: Then we have two places to update for model name... What happens if they go out of sync?

  - models
  - model table

File: domain/schema/controller_test.go (hunk 1 above)
Comment (SimonRichardson) [reply domain/schema/controller_test.go:79]: I've left this for now. I want to group schema tables in the tests as it's easier to reason about why they exist from a test perspective.

File: domain/schema/controller.go [hunk 3]
```diff
@@ -272,3 +273,15 @@ BEGIN
 END;
 `[1:]
 }
+
+func modelListSchema() string {
+	return `
+CREATE TABLE model_list (
+    uuid  TEXT NOT NULL
```
Comment (manadart) [inline domain/schema/controller.go:280]: Just go with `TEXT PRIMARY KEY` and drop the custom index.

**Review by manadart — APPROVED** [review]
Thanks for the changes.

**Pull request merged**